src/services/swap_service.ts

```diff
@@ -332,21 +332,32 @@ export class SwapService {
         // using eth_gasEstimate
         // If an error occurs we attempt to provide a better message then "Transaction Reverted"
         if (takerAddress && !skipValidation && canEstimateGas) {
-            let estimateGasCallResult = await this._estimateGasOrThrowRevertErrorAsync({
-                to,
-                data,
-                from: takerAddress,
-                value,
-                gasPrice,
-            });
-            // Add any underterministic gas overhead the encoded transaction has detected
-            estimateGasCallResult = estimateGasCallResult.plus(gasOverhead);
-            // Take the max of the faux estimate or the real estimate
-            conservativeBestCaseGasEstimate = BigNumber.max(
-                // Add a little buffer to eth_estimateGas as it is not always correct
-                estimateGasCallResult.times(GAS_LIMIT_BUFFER_MULTIPLIER).integerValue(),
-                conservativeBestCaseGasEstimate,
-            );
+            try {
+                let estimateGasCallResult = await this._estimateGasOrThrowRevertErrorAsync({
+                    to,
+                    data,
+                    from: takerAddress,
+                    value,
+                    gasPrice,
+                });
+                // Add any underterministic gas overhead the encoded transaction has detected
+                estimateGasCallResult = estimateGasCallResult.plus(gasOverhead);
+                // Take the max of the faux estimate or the real estimate
+                conservativeBestCaseGasEstimate = BigNumber.max(
+                    // Add a little buffer to eth_estimateGas as it is not always correct
+                    estimateGasCallResult.times(GAS_LIMIT_BUFFER_MULTIPLIER).integerValue(),
+                    conservativeBestCaseGasEstimate,
+                );
+            } catch (error) {
+                if (isFirmQuote) {
+                    // On firm quotes, when skipValidation=false, we want to raise an error
+                    throw error;
+                }
+                logger.warn(
+                    { takerAddress, data, value, gasPrice, error: error?.message },
+                    'Unable to use eth_estimateGas. Falling back to faux estimate',
+                );
```

Comment:
This is great, but I'm wondering if we are getting some different unintended consequences. For `/price` it's great that we can fall back to the gas schedule if we can't use `eth_estimateGas`. 

But I'm fairly sure this would also affect `/quote`, which means that we may give inaccurate gas estimations even tho the taker address was supplied, or worse we give out quotes that are guaranteed to revert due to missing balance/allowance etc.

Reply:
Ahh - for quote, we will rethrow the error, so the behavior should be the same as before. Our tests actually caught this and got me to add this above:

```
                if (isFirmQuote) {
                    // On firm quotes, when skipValidation=false, we want to raise an error
                    throw error;
                }
```

Reply:
Currently adding a test around `/price` so we'll have some certainty that both `/price` and `/quote` will work as intended

Reply:
Correct me if I'm wrong, but I think with the current logic it would not throw for `/quote` requests that do not use RFQ, or am I missing something?

Reply:
Ahh, I think you were right

I did just update the code to fix some tests and in doing so, I think it also happens to fix the case you were asking about (validation on `/quote` when no api key is present)

I think the confusion here was around what the boolean `isFirmQuote` actually represented. I've made it much simpler by just using the `endpoint` to distinguish if we're looking at a quote or a price. Let me know if the new version checks out to you!

Reply:
I think it's correctly throwing for quotes now but with `const canEstimateGas = isQuote || !isNativeIncluded;` aren't you back to skipping gas estimations for all `/price` requests? 😅